pkg/utils/kube/uniqueprotomodels.go

```diff
@@ -0,0 +1,184 @@
+
+So we have to do some work to make the OpenAPI document palatable to the GVKParser. We have to remove the duplicate
+definitions. Specifically, we take the first one and log a warning for each subsequent definition with the same GVK.
+
+In practice, this probably generally appears when a common aggregated API was built at a time significantly before the
+current Kubernetes version. The most common case is that the metrics server is built against an older version of the
+Kubernetes libraries, using old versions of the structs. When the metrics server is updated to use the latest version of
+the Kubernetes libraries, the problems go away, because the aggregated API and Kubernetes agree about the shape of the
+struct.
+
+Using the first encountered definition is imperfect and could result in unreliable diffs. But it's better than
+constructing completely-wrong diffs due to the lack of a GVKParser.
+*/
+
+// uniqueModels is a model provider that ensures that no two models share the same gvk. Use newUniqueModels to
+// initialize it and enforce uniqueness.
+type uniqueModels struct {
+	models map[string]proto.Schema
+}
+
+// LookupModel is public through the interface of Models. It
+// returns a visitable schema from the given model name.
+// Copied verbatim from here: https://github.com/kubernetes/kube-openapi/blob/b456828f718bab62dc3013d192665eb3d17f8fe9/pkg/util/proto/document.go#L322-L326
+func (d *uniqueModels) LookupModel(model string) proto.Schema {
+	return d.models[model]
+}
+
+// Copied verbatim from here: https://github.com/kubernetes/kube-openapi/blob/b456828f718bab62dc3013d192665eb3d17f8fe9/pkg/util/proto/document.go#L328-L337
+func (d *uniqueModels) ListModels() []string {
+	models := []string{}
+
+	for model := range d.models {
+		models = append(models, model)
+	}
+
+	sort.Strings(models)
+	return models
+}
+
+// newUniqueModels returns a new uniqueModels instance and a list of warnings for models that share the same gvk.
+func newUniqueModels(models proto.Models) (proto.Models, []schema.GroupVersionKind) {
+	var taintedGVKs []schema.GroupVersionKind
+	gvks := map[schema.GroupVersionKind]string{}
+	um := &uniqueModels{models: map[string]proto.Schema{}}
+	for _, modelName := range models.ListModels() {
+		model := models.LookupModel(modelName)
+		if model == nil {
+			panic(fmt.Sprintf("ListModels returns a model that can't be looked-up for: %v", modelName))
+		}
+		gvkList := parseGroupVersionKind(model)
+		gvk, wouldError := wouldTriggerDuplicateError(model, gvks)
+		if !wouldError {
+			um.models[modelName] = model
+
+			// Add GVKs to the map, so we can detect a duplicate GVK later.
+			for _, gvk := range gvkList {
+				if len(gvk.Kind) > 0 {
+					gvks[gvk] = modelName
+				}
+			}
+		} else {
+			taintedGVKs = append(taintedGVKs, gvk)
+		}
+	}
+	return um, taintedGVKs
+}
+
+func wouldTriggerDuplicateError(model proto.Schema, gvks map[schema.GroupVersionKind]string) (schema.GroupVersionKind, bool) {
```

Comment:
I would add a brief godoc to clarify what this functions does and what the return type means. I would also change the function name to something related to what the function does. Suggestion `findDuplicatedGVKs`, `hasDuplicatedGVKs`.